File: apps/desktop/src-tauri/src/lex.rs
// Lexical state -> plain text extraction (server-side mirror of the renderer).
//
// Lossy on purpose - feeds FTS5 and the plain-text export, NOT round-trip
// rendering.

use serde_json::Value;

#[derive(Debug, Clone)]
pub struct ExtractedBlock {
    pub kind: String,
    pub text: String,
}

pub fn extract_blocks(content_json: &str) -> Vec<ExtractedBlock> {
    let v: Value = match serde_json::from_str(content_json) {
        Ok(v) => v,
        Err(_) => return vec![],
    };
    let root = v.get("root").cloned().unwrap_or(v);
    let mut out = Vec::new();
    walk(&root, &mut out);
    out
}

fn walk(node: &Value, out: &mut Vec<ExtractedBlock>) {
    if let Some(typ) = node.get("type").and_then(|t| t.as_str()) {
        let scriptz_kinds = [
            "scriptz-action",
            "scriptz-character",
            "scriptz-dialog",
            "scriptz-parenthetical",
            "scriptz-camera",
            "scriptz-caption",
            "scriptz-sfx",
        ];
        if scriptz_kinds.contains(&typ) {
            let text = collect_text(node);
            out.push(ExtractedBlock {
                kind: typ.to_string(),
                text,
            });
            return;
        }
    }
    if let Some(children) = node.get("children").and_then(|c| c.as_array()) {
        for child in children {
            walk(child, out);
        }
    }
}

fn collect_text(node: &Value) -> String {
    let mut s = String::new();
    collect_text_into(node, &mut s);
    s
}

fn collect_text_into(node: &Value, s: &mut String) {
    if let Some(t) = node.get("type").and_then(|v| v.as_str()) {
        if t == "linebreak" {
            s.push('\n');
            return;
        }
        if t == "text" {
            if let Some(text) = node.get("text").and_then(|v| v.as_str()) {
                s.push_str(text);
            }
            return;
        }
    }
    if let Some(children) = node.get("children").and_then(|c| c.as_array()) {
        for child in children {
            collect_text_into(child, s);
        }
    }
}

pub fn extract_plain_text(content_json: &str) -> String {
    let blocks = extract_blocks(content_json);
    let mut out = String::new();
    for b in blocks {
        if !out.is_empty() {
            out.push_str("\n\n");
        }
        match b.kind.as_str() {
            "scriptz-character" => {
                out.push_str(&b.text.to_uppercase());
            }
            "scriptz-parenthetical" => {
                let trimmed = b.text.trim_matches(|c| c == '(' || c == ')');
                out.push('(');
                out.push_str(trimmed);
                out.push(')');
            }
            "scriptz-sfx" => {
                if !b.text.to_uppercase().starts_with("SFX:") {
                    out.push_str("SFX: ");
                }
                out.push_str(&b.text);
            }
            _ => out.push_str(&b.text),
        }
    }
    out
}

pub fn extract_teleprompter_text(content_json: &str) -> String {
    // Plain-Text export per spec: only Character, Dialog, Parenthetical.
    let blocks = extract_blocks(content_json);
    let mut out = String::new();
    for b in blocks {
        match b.kind.as_str() {
            "scriptz-character" => {
                if !out.is_empty() {
                    out.push_str("\n\n");
                }
                out.push_str(&b.text.to_uppercase());
            }
            "scriptz-dialog" => {
                if !out.is_empty() {
                    out.push('\n');
                }
                out.push_str(&b.text);
            }
            "scriptz-parenthetical" => {
                if !out.is_empty() {
                    out.push('\n');
                }
                let trimmed = b.text.trim_matches(|c| c == '(' || c == ')');
                out.push('(');
                out.push_str(trimmed);
                out.push(')');
            }
            _ => {}
        }
    }
    out
}

/// Word-token set for similarity comparison. Lowercased, ASCII-stripped of
/// punctuation, ignores tokens shorter than 3 chars (drops "im", "am", "und"
/// etc. that would dominate a Jaccard score). NOT meant for FTS — that's what
/// `extract_plain_text` feeds.
pub fn word_token_set(text: &str) -> std::collections::HashSet<String> {
    let mut out = std::collections::HashSet::new();
    for raw in text.split(|c: char| !c.is_alphanumeric()) {
        let token: String = raw.chars().flat_map(|c| c.to_lowercase()).collect();
        if token.chars().count() >= 3 {
            out.insert(token);
        }
    }
    out
}

/// Jaccard similarity between two word-token sets, in [0.0, 1.0].
/// Two empty sets are defined as fully similar (1.0) — no content, no diff.
pub fn jaccard_similarity(
    a: &std::collections::HashSet<String>,
    b: &std::collections::HashSet<String>,
) -> f32 {
    if a.is_empty() && b.is_empty() {
        return 1.0;
    }
    let inter = a.intersection(b).count() as f32;
    let union = a.union(b).count() as f32;
    if union == 0.0 { 1.0 } else { inter / union }
}

// extract_character_names retired in Migration Phase 7d — only the
// Rust-side reconcile used it, and the reconcile now runs in TS via
// src/lib/lex.ts (see Phase 1 byte-equivalence proof).

File: packages/core/lib/lex.ts
// Lexical state -> plain text extraction (server-side mirror of the renderer).
//
// Lossy on purpose - feeds FTS5 and the plain-text export, NOT round-trip
// rendering.
//
// Mirrors src-tauri/src/lex.rs byte-for-byte during the Rust -> TS migration.
// Both modules coexist until Phase 11 retires the Rust copy.

/** A contiguous text fragment with uniform inline formatting.
 *  Mirrors Lexical's per-TextNode `format` bitfield: bit 0 = bold,
 *  bit 1 = italic, bit 3 = underline. Linebreaks are flattened into
 *  `\n` inside `text`. */
export interface TextRun {
  text: string;
  bold: boolean;
  italic: boolean;
  underline: boolean;
}

export interface ExtractedBlock {
  kind: string;
  text: string;
  /** Per-character formatting, preserved so the PDF can render bold/italic
   *  spans inside a dialog without losing the character tint. Always present;
   *  for unformatted text it contains a single run with all three flags false. */
  runs: TextRun[];
}

const FORMAT_BOLD = 1;
const FORMAT_ITALIC = 2;
const FORMAT_UNDERLINE = 8;

const SCRIPTZ_KINDS = new Set([
  "scriptz-action",
  "scriptz-character",
  "scriptz-dialog",
  "scriptz-parenthetical",
  "scriptz-camera",
  "scriptz-caption",
  "scriptz-sfx",
]);

export function extractBlocks(contentJson: string): ExtractedBlock[] {
  let v: unknown;
  try {
    v = JSON.parse(contentJson);
  } catch {
    return [];
  }
  const root = isObject(v) && "root" in v ? (v as Record<string, unknown>).root : v;
  const out: ExtractedBlock[] = [];
  walk(root, out);
  return out;
}

function walk(node: unknown, out: ExtractedBlock[]): void {
  if (!isObject(node)) return;
  const typ = node.type;
  if (typeof typ === "string" && SCRIPTZ_KINDS.has(typ)) {
    const runs = collectRuns(node);
    const text = runs.map((r) => r.text).join("");
    out.push({ kind: typ, text, runs: mergeRuns(runs) });
    return;
  }
  const children = node.children;
  if (Array.isArray(children)) {
    for (const child of children) walk(child, out);
  }
}

function collectRuns(node: unknown): TextRun[] {
  const buf: TextRun[] = [];
  collectRunsInto(node, buf);
  if (buf.length === 0) buf.push({ text: "", bold: false, italic: false, underline: false });
  return buf;
}

function collectRunsInto(node: unknown, buf: TextRun[]): void {
  if (!isObject(node)) return;
  const t = node.type;
  if (typeof t === "string") {
    if (t === "linebreak") {
      buf.push({ text: "\n", bold: false, italic: false, underline: false });
      return;
    }
    if (t === "text") {
      const text = node.text;
      if (typeof text === "string") {
        const fmt = typeof node.format === "number" ? node.format : 0;
        buf.push({
          text,
          bold: (fmt & FORMAT_BOLD) !== 0,
          italic: (fmt & FORMAT_ITALIC) !== 0,
          underline: (fmt & FORMAT_UNDERLINE) !== 0,
        });
      }
      return;
    }
  }
  const children = node.children;
  if (Array.isArray(children)) {
    for (const child of children) collectRunsInto(child, buf);
  }
}

/** Coalesce adjacent runs with identical formatting so downstream code
 *  doesn't draw a font switch for what is logically one span. */
function mergeRuns(runs: TextRun[]): TextRun[] {
  if (runs.length <= 1) return runs;
  const out: TextRun[] = [{ ...runs[0] }];
  for (let i = 1; i < runs.length; i++) {
    const last = out[out.length - 1];
    const cur = runs[i];
    if (last.bold === cur.bold && last.italic === cur.italic && last.underline === cur.underline) {
      last.text += cur.text;
    } else {
      out.push({ ...cur });
    }
  }
  return out;
}

export function extractPlainText(contentJson: string): string {
  const blocks = extractBlocks(contentJson);
  let out = "";
  for (const b of blocks) {
    if (out.length > 0) out += "\n\n";
    switch (b.kind) {
      case "scriptz-character":
        out += b.text.toUpperCase();
        break;
      case "scriptz-parenthetical": {
        const trimmed = trimParens(b.text);
        out += "(" + trimmed + ")";
        break;
      }
      case "scriptz-sfx":
        if (!b.text.toUpperCase().startsWith("SFX:")) out += "SFX: ";
        out += b.text;
        break;
      default:
        out += b.text;
    }
  }
  return out;
}

export function extractTeleprompterText(contentJson: string): string {
  // Plain-Text export per spec: only Character, Dialog, Parenthetical.
  const blocks = extractBlocks(contentJson);
  let out = "";
  for (const b of blocks) {
    switch (b.kind) {
      case "scriptz-character":
        if (out.length > 0) out += "\n\n";
        out += b.text.toUpperCase();
        break;
      case "scriptz-dialog":
        if (out.length > 0) out += "\n";
        out += b.text;
        break;
      case "scriptz-parenthetical": {
        if (out.length > 0) out += "\n";
        const trimmed = trimParens(b.text);
        out += "(" + trimmed + ")";
        break;
      }
      default:
        break;
    }
  }
  return out;
}

/**
 * Word-token set for similarity comparison. Lowercased, ASCII-stripped of
 * punctuation, ignores tokens shorter than 3 chars (drops "im", "am", "und"
 * etc. that would dominate a Jaccard score). NOT meant for FTS - that's what
 * `extractPlainText` feeds.
 */
export function wordTokenSet(text: string): Set<string> {
  const out = new Set<string>();
  // Mirrors Rust's `c.is_alphanumeric()` (Unicode Alphabetic | Numeric).
  const parts = text.split(/[^\p{Alphabetic}\p{N}]+/u);
  for (const raw of parts) {
    if (raw.length === 0) continue;
    const token = raw.toLowerCase();
    // Count Unicode scalar values, not UTF-16 code units, to mirror
    // Rust's `chars().count() >= 3`.
    if (codePointCount(token) >= 3) out.add(token);
  }
  return out;
}

/**
 * Jaccard similarity between two word-token sets, in [0.0, 1.0].
 * Two empty sets are defined as fully similar (1.0) - no content, no diff.
 */
export function jaccardSimilarity(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 && b.size === 0) return 1.0;
  let inter = 0;
  for (const t of a) if (b.has(t)) inter++;
  const union = a.size + b.size - inter;
  if (union === 0) return 1.0;
  return inter / union;
}

/** Dialog words per character - same algorithm as in the EditorRail
 *  cast tab: words from `scriptz-dialog` blocks are assigned to the
 *  most recently preceding `scriptz-character` block. Characters
 *  without dialog have an entry of `0` (so callers know
 *  they exist). Keys are UPPERCASE because character names
 *  match case-insensitively app-wide.
 *
 *  Single source of truth for this calculation - see `scripts.ts`
 *  (for `characters_meta.share` on save) and `EditorRail.tsx`
 *  (for the cast tab). */
export function dialogWordsByCharacter(contentJson: string): Record<string, number> {
  const out: Record<string, number> = {};
  let last: string | null = null;
  for (const b of extractBlocks(contentJson)) {
    if (b.kind === "scriptz-character") {
      last = b.text.trim().toUpperCase();
      if (last && !(last in out)) out[last] = 0;
    } else if (b.kind === "scriptz-dialog" && last) {
      const w = b.text.trim().split(/\s+/).filter(Boolean).length;
      out[last] = (out[last] ?? 0) + w;
    }
  }
  return out;
}

/** Collect unique character names (uppercase) used in the script's content. */
export function extractCharacterNames(contentJson: string): string[] {
  const blocks = extractBlocks(contentJson);
  const seen = new Set<string>();
  const out: string[] = [];
  for (const b of blocks) {
    if (b.kind !== "scriptz-character") continue;
    const name = b.text.trim().toUpperCase();
    if (name.length === 0) continue;
    if (!seen.has(name)) {
      seen.add(name);
      out.push(name);
    }
  }
  return out;
}

// ---------- helpers ----------

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

// Mirrors `b.text.trim_matches(|c| c == '(' || c == ')')`: strip ALL
// leading/trailing parens (not just one pair).
function trimParens(s: string): string {
  let start = 0;
  let end = s.length;
  while (start < end && (s[start] === "(" || s[start] === ")")) start++;
  while (end > start && (s[end - 1] === "(" || s[end - 1] === ")")) end--;
  return s.slice(start, end);
}

function codePointCount(s: string): number {
  let n = 0;
  for (const _ of s) n++;
  return n;
}
